### ISSUE TITLE
CBLOF predict error

### ISSUE THREAD

**Issue by GiovannaR** [yzhao062/pyod#41]
Hi,
When I try to use CBLOF to predict one or two or any short number of samples, sometimes it fails like in the example above:

```
clf_name = 'CBLOF'
clf = CBLOF(alpha=0.7, beta=2, check_estimator=False, n_clusters=6)
clf.fit(a[0:336])
print([a[338]])
clf.predict([a[338]])
```
Output:

```
[array([0.21751617])]
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-173-5342692feffe> in <module>()
      3 clf.fit(a[0:336])
      4 
----> 5 clf.predict([a[338]])

/usr/local/lib/python3.5/dist-packages/pyod/models/base.py in predict(self, X)
    125         check_is_fitted(self, ['decision_scores_', 'threshold_', 'labels_'])
    126 
--> 127         pred_score = self.decision_function(X)
    128         return (pred_score > self.threshold_).astype('int').ravel()
    129 

/usr/local/lib/python3.5/dist-packages/pyod/models/cblof.py in decision_function(self, X)
    179         X = check_array(X)
    180         labels = self.clustering_estimator_.predict(X)
--> 181         return self._decision_function(X, labels)
    182 
    183     def _validate_estimator(self, default=None):

/usr/local/lib/python3.5/dist-packages/pyod/models/cblof.py in _decision_function(self, X, labels)
    281 
    282         scores[large_indices] = pairwise_distances_no_broadcast(
--> 283             X[large_indices, :], large_centers)
    284 
    285         if self.use_weights:

/usr/local/lib/python3.5/dist-packages/pyod/utils/stat_models.py in pairwise_distances_no_broadcast(X, Y)
     36     :rtype: array of shape (n_samples,)
     37     """
---> 38     X = check_array(X)
     39     Y = check_array(Y)
     40     assert_allclose(X.shape, Y.shape)

/usr/local/lib/python3.5/dist-packages/sklearn/utils/validation.py in check_array(array, accept_sparse, accept_large_sparse, dtype, order, copy, force_all_finite, ensure_2d, allow_nd, ensure_min_samples, ensure_min_features, warn_on_dtype, estimator)
    580                              " minimum of %d is required%s."
    581                              % (n_samples, shape_repr, ensure_min_samples,
--> 582                                 context))
    583 
    584     if ensure_min_features > 0 and array.ndim == 2:

ValueError: Found array with 0 sample(s) (shape=(0, 1)) while a minimum of 1 is required.
```
But when I try to predict ensuring that one of them is not an anomaly, then it works in all the cases:

```
pred = clf.predict([clf.cluster_centers_[clf.large_cluster_labels_[0]],a[338]])
print (pred)
```

Output:

```
[0 1]
```

Thanks for your help

**Comment by yzhao062** [comment]
Hey,

It fails at array shape check. If it is a public dataset, would you mind sharing it?
If not, is this also failing with the artificial data generated by generate_data() in pyod?

Just try to reproduce first :)
Thanks,
Yue

**Comment by GiovannaR** [comment]
Hey,

It also happens with the artificial data

Example:
```
test = generate_data(train_only=True)
clf_name = 'CBLOF'
clf = CBLOF(alpha=0.7, beta=2, check_estimator=False, n_clusters=6)
clf.fit(test[0][0:336])
clf.predict([test[0][338]])
```
Output:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-29-5b9cd080d1de> in <module>()
----> 1 clf.predict([test[0][338]])

/usr/local/lib/python3.5/dist-packages/pyod/models/base.py in predict(self, X)
    125         check_is_fitted(self, ['decision_scores_', 'threshold_', 'labels_'])
    126 
--> 127         pred_score = self.decision_function(X)
    128         return (pred_score > self.threshold_).astype('int').ravel()
    129 

/usr/local/lib/python3.5/dist-packages/pyod/models/cblof.py in decision_function(self, X)
    179         X = check_array(X)
    180         labels = self.clustering_estimator_.predict(X)
--> 181         return self._decision_function(X, labels)
    182 
    183     def _validate_estimator(self, default=None):

/usr/local/lib/python3.5/dist-packages/pyod/models/cblof.py in _decision_function(self, X, labels)
    281 
    282         scores[large_indices] = pairwise_distances_no_broadcast(
--> 283             X[large_indices, :], large_centers)
    284 
    285         if self.use_weights:

/usr/local/lib/python3.5/dist-packages/pyod/utils/stat_models.py in pairwise_distances_no_broadcast(X, Y)
     36     :rtype: array of shape (n_samples,)
     37     """
---> 38     X = check_array(X)
     39     Y = check_array(Y)
     40     assert_allclose(X.shape, Y.shape)

/usr/local/lib/python3.5/dist-packages/sklearn/utils/validation.py in check_array(array, accept_sparse, accept_large_sparse, dtype, order, copy, force_all_finite, ensure_2d, allow_nd, ensure_min_samples, ensure_min_features, warn_on_dtype, estimator)
    580                              " minimum of %d is required%s."
    581                              % (n_samples, shape_repr, ensure_min_samples,
--> 582                                 context))
    583 
    584     if ensure_min_features > 0 and array.ndim == 2:

ValueError: Found array with 0 sample(s) (shape=(0, 2)) while a minimum of 1 is required.
```
But when I try to predict ensuring that one of them is not an anomaly, then it works:

```
clf.predict([clf.cluster_centers_[clf.large_cluster_labels_[0]],test[0][338]])
```
Output:
```
array([0, 0])
```
Thanks for your help,
Giovanna

**Comment by yzhao062** [comment]
Roger that! Will investigate and update soon :)
Thanks! Yue

**Comment by yzhao062** [comment]
Hi. Thanks for reporting this bug! It is fixed in #44 & pre-lease v.0.6.6b1 (https://pypi.org/project/pyod/0.6.6b1/) and planned to be released in v0.6.6 later this week (likely to be this Sat). If you would like to try the pre-release, then 

> pip install pyod==0.6.6b1

Otherwise, please wait for 4 more days for an official release...

The cause of the bug is when only a single sample passed in, it could only be clustered into either large or small clusters, and it will fail with either the calculation for small or large clusters. We added a condition check to prevent this from happening.

Meanwhile, when passing in a single sample, please use **.reshape(1,-1)** to reduce the confusion :)

**Comment by yzhao062** [comment]
This bug is fixed in [v0.6.6.](https://pypi.org/project/pyod/).
Please update your local installation with pip :)
Feel free to reopen the issue if it is not working on your end. Thanks.
